hypershift-operator/controllers/awsendpointservice/awsendpointservice_controller.go

```diff
@@ -0,0 +1,201 @@
+	// fetch the AWSEndpointService
+	awsEndpointService := &hyperv1.AWSEndpointService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+	err := r.Get(ctx, client.ObjectKeyFromObject(awsEndpointService), awsEndpointService)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get resource: %w", err)
+	}
+
+	// Return early if deleted
+	if !awsEndpointService.DeletionTimestamp.IsZero() {
+		completed, err := r.delete(ctx, awsEndpointService)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete resource: %w", err)
+		}
+		if !completed {
+			return ctrl.Result{RequeueAfter: endpointServiceDeletionRequeueDuration}, nil
+		}
+		if controllerutil.ContainsFinalizer(awsEndpointService, finalizer) {
+			controllerutil.RemoveFinalizer(awsEndpointService, finalizer)
+			if err := r.Update(ctx, awsEndpointService); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Ensure the awsEndpointService has a finalizer for cleanup
+	if !controllerutil.ContainsFinalizer(awsEndpointService, finalizer) {
+		controllerutil.AddFinalizer(awsEndpointService, finalizer)
+		if err := r.Update(ctx, awsEndpointService); err != nil {
+			if apierrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	// Reconcile the AWSEndpointService
+	err = reconcileAWSEndpointService(ctx, awsEndpointService, r.ec2Client, r.elbv2Client)
+	if err != nil {
+		meta.SetStatusCondition(&awsEndpointService.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.AWSEndpointServiceAvailable),
+			Status:  metav1.ConditionFalse,
+			Reason:  hyperv1.AWSErrorReason,
+			Message: err.Error(),
+		})
+	} else {
+		meta.SetStatusCondition(&awsEndpointService.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.AWSEndpointServiceAvailable),
+			Status:  metav1.ConditionTrue,
+			Reason:  hyperv1.AWSSuccessReason,
+			Message: "",
+		})
+	}
+	r.Status().Update(ctx, awsEndpointService)
+	return ctrl.Result{}, err
+}
+
+func reconcileAWSEndpointService(ctx context.Context, awsEndpointService *hyperv1.AWSEndpointService, ec2Client ec2iface.EC2API, elbv2Client elbv2iface.ELBV2API) error {
+	serviceName := awsEndpointService.Status.EndpointServiceName
+	if len(serviceName) != 0 {
+		// check if Endpoint Service exists in AWS
+		output, err := ec2Client.DescribeVpcEndpointServiceConfigurationsWithContext(ctx, &ec2.DescribeVpcEndpointServiceConfigurationsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("service-name"),
+					Values: []*string{aws.String(serviceName)},
+				},
+			},
+		})
+		if err != nil {
+			return err
+		}
+		if len(output.ServiceConfigurations) == 0 {
```

Comment:
Shouldn't we remove it from the status if that happens so that it gets recreated in the next iteration?

Reply:
Good idea :+1: